docs/plugins.md

```diff
@@ -1,5 +1,202 @@
+
+ 1. Plugins that modify the request should be at the beginning (like the `AuthenticationPlugin` or the `CookiePlugin`)
+ 2. Plugins which intervene in the workflow should be in the "middle" (like the `RetryPlugin` or the `RedirectPlugin`)
+ 3. Plugins which log and a debug information about a call should be last (like the `LoggerPlugin` or the `HistoryPlugin`)
+
+However there may be exception in this rule, if, for security reason, you don't want to log the authentication header, it will be better to put
+this plugin after the `LoggerPlugin` one.
+
+## Implementing your own Plugin
+
+When writing your own Plugin, you need to be aware that the `PluginClient` is async first. It means that every plugin must
+be written by respecting the `HttpAsyncClient` contract and use `Promise` as the return.
+
+Each plugin must implement the `Http\Client\Plugin\Plugin` interface.
+
+This interface defined the `handleRequest` method which allow to modify behavior of the call:
+
+```php
+/**
+ * handle the request and return the response coming from the next callable
+ *
+ * @param RequestInterface $request Request to use
+ * @param callable         $next    Callback to call to have the request, it muse have the request as it first argument
+ * @param callable         $first   First element in the plugin chain, used to to restart a request from the beginning
+ *
+ * @return Promise
+ */
+public function handleRequest(RequestInterface $request, callable $next, callable $first);
+```
+
+The `$request` is the one created by the client, you can transform it, or do what you like with it. Always be aware that
+the request is immutable.
+
+```
+public function handleRequest(RequestInterface $request, callable $next, callable $first)
+{
+    $newRequest = $request->withHeader('MyHeader', 'MyValue');
+}
+```
+
+The `$next` callable is the next plugin in the execution chain. When you need to call it, you must pass the `$request` 
+as the first argument of this callable.
+
+```
+public function handleRequest(RequestInterface $request, callable $next, callable $first)
+{
+    $newRequest = $request->withHeader('MyHeader', 'MyValue');
+    
+    return $next($request);
+}
+```
+
+
+The `$first` callable is the first plugin in the execution. It allows you to completely reboot the execution chain, or sends 
+other request if needed while still using all the plugins defined. Like the `$next` callable, you must pass the `$request` 
+as the first argument of this callable.
+
+```
+public function handleRequest(RequestInterface $request, callable $next, callable $first)
+{
+    if ($someCondition) {
+        $newRequest = new Request();
+        $promise    = $first($newRequest);
+        
+        // Use the promise do some jobs ...
+    }
+    
+    return $next($request);
+}
+```
+
+In this example the condition is not superfluous, you need to have some way to not calling the $first callable each time or
+you will end up with a infinite execution loop.
```

Comment:
maybe have an option to define how many restarts should be allowed?
we could also just accept a plugin for that, but thats very implicit and
not happening by default. and imho this is something that should happen
by default to protect people not aware of the risk. endless redirects
are pretty simple to create if the server is doing something stupid
(like redirect to the same url you are already on)